docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,507 @@
+nuanced. In this case, the optional behavior is never allowed if the document's
+type has changed. That makes for a mandatory test, one that verifies browsers
+don't return the same result when the document's type changes.
+
+## Exercising Restraint
+
+When writing conformance tests, choosing what *not* to test is sometimes just
+as hard as finding what needs testing.
+
+### Don't dive too deep
+
+Algorithms are usually composed of many other algorithms which themselves are
+defined in terms of still more algorithms. It can be intimidating to consider
+exhaustively testing one of those "nested" algorithms, especially when they are
+shared by many different APIs.
+
+In general, you should start by writing only "surface tests" for the nested
+algorithms. That means only verifying that they exhibit the basic behavior you
+are expecting.
+
+It's definitely important to test exhaustively, but it's just as important to
+do so in a structured way. Reach out to the test suite's maintainers to learn
+if and how they have already tested those algorithms. In many cases, it's
+acceptable to test them in just one place (and maybe through a different API
+entirely), and rely only on surface-level testing everywhere else. While it's
+always possible for more tests to uncover new bugs, the chances may be slim.
+The time we spend writing tests is highly valuable, so we have to be efficient!
+
+*Example:* The following algorithm from [the DOM
+standard](https://dom.spec.whatwg.org/) powers
+[`document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector):
+
+> To **scope-match a selectors string** *selectors* against a *node*, run these
+> steps:
+>
+> 1. Let *s* be the result of [parse a
+>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
+>    *selectors*.
+> 2. If *s* is failure, then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
+>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 3. Return the result of [match a selector against a
+>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
+>    with *s* and *node*'s
+>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
+>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.
+
+As described earlier in this guide, we'd certainly want to test the branch
+regarding the parsing failure. However, there are many ways a string might fail
+to parse--should we verify them all in the tests for `document.querySelector`?
+What about `document.querySelectorAll`? Should we test them all there, too?
+
+The answers depend on the current state of the test suite: whether or not tests
+for selector parsing exist and where they are located. That's why it's best to
+confer with the people who are maintaining the tests.
+
+### Avoid excessive breadth
```

Comment:
So, I'm not sure where we are on this section atm, but I think currently it reads too negatively towards dynamically generated tests. Yes, this is ironic since I previously argued that dynamic generation of tests is bad, but I think there's a balance and this text strays too far towards 'dont do it'

I would like to see it balanced slightly by acknowledging that this is a complex area, and sometimes it helps and sometimes it doesn't. I would be ok with pushing people away from it, given that this is a guide, but I just don't want to indoctrinate people into thinking never do it.

Reply:
I admit to being more strongly opposed to the practice than most, but I also think that by introducing a consideration without offering some advise, we may make contributors feel more doubtful than empowered. It's sort of like telling someone, "the hardest part about deep-sea fishing is making sure the robotic sharks don't destroy your boat," and moving on to discuss the intricacies of lure selection.

Really, I think we're already risking that. All but one sentence is couched in some way ("can", "often", "may"). I'm surprised that the latest revision feels too forceful. The word "sometimes" already appears in the conclusion:

"Although the risks of dynamic test generation may be tolerable in some specific cases, it’s sometimes best to select the most interesting edge cases and move on."

The risk of indoctrination from a statement like that seems pretty low.

Reply:
The "may be tolerable in some specific cases" wording, which still remains, conveys quite strongly that this usually a very bad idea and that you'd better default to not writing for loops in your tests.

Since I now have @stephenmcgruer on my side of the argument, I'm happy to push back on this again. This, I think, is *just fine*, and better than unrolling it if the test body is longer:
```js
for (const tagName of ['audio', 'video']) {
  for (const volume of [0, 0.5, 1]) {
    test(() => {
      const element = document.createElement(tagName);
      element.volume = volume;
      assert_something_interesting();
    }, `${tagName} volume ${volume}`);
  }
}
```

Reply:
Also, maybe advice around this makes more sense somewhere else? This guide is mostly about deciding *what* and *how much* to test, not whether to dynamically generate the tests or not.